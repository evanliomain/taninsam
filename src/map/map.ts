/**
 * @module array=>array
 */
import { Function, Iteratee, Links } from '../@types';
/**
 * This method creates a new array
 * with the results of calling a provided function
 * on every element in the calling array.
 * See the [native map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) for more informations.
 * @param iteratee The iteratee invoked per element.
 * @return the function to apply on the array to transform
 * @example
 * ```
 * map<number, number>(x => 1 + x))([1, 2, 3, 4, 5]) // [2, 3, 4, 5, 6]
 * ```
 * @example Using the chain
 * ```
 * chain([1, 2, 3, 4, 5])
 *   .chain(map<number, number>(x => 1 + x)))
 *   .value() // [2, 3, 4, 5, 6]
 * ```
 */
export function map<T, U>(
  iteratee: Iteratee<T, U>,
): Function<ReadonlyArray<T>, ReadonlyArray<U>> {
  return (array: ReadonlyArray<T>, links?: Links) =>
    array.map((value: T, index: number, originalArray: ReadonlyArray<T>) =>
      iteratee(value, index, originalArray, links),
    );
}
