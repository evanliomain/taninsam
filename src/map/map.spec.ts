import { map } from './map';

describe('map function', () => {
  const input = [1, 2, 3, 4, 5];
  const incr = x => 1 + x;

  test('is a pure function', () => {
    map(incr)(input);
    expect(input).toMatchSnapshot();
  });

  test('[1, 2, 3, 4, 5] |> map(incr) === [2, 3, 4, 5, 6]', () => {
    expect(map(incr)(input)).toMatchSnapshot();
  });

  test('[ |> map(incr) === []', () => {
    expect(map(incr)([])).toMatchSnapshot();
  });

  test('call the iteree callback with correct inputs', () => {
    const cb = jest.fn((x: number) => 1 + x);
    const link = { fake: 'fake' };
    map<number, number>(cb)(input, link);
    input.forEach((i, index) => {
      expect(cb.mock.calls[index][0]).toBe(i);
      expect(cb.mock.calls[index][1]).toBe(index);
      expect(cb.mock.calls[index][2]).toEqual(input);
      expect(cb.mock.calls[index][3]).toBe(link);
    });
  });
});
