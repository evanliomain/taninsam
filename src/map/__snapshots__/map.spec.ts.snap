// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map function [] |> map(incr) === [] 1`] = `Array []`;

exports[`map function [1, 2, 3, 4, 5] |> map(incr) === [2, 3, 4, 5, 6] 1`] = `
Array [
  2,
  3,
  4,
  5,
  6,
]
`;

exports[`map function is a pure function 1`] = `
Array [
  1,
  2,
  3,
  4,
  5,
]
`;
